COM-392 Appts: Start and end time not mandatory

prashantsable1991 opened the pull request:
Removed required from start and end time. Added logic to save an appointment when start time and end time, not provided.

lauravignoli commented on ui/app/appointments/controllers/manage/calendar/appointmentsCalendarViewController.js:
```diff
@@ -117,14 +117,15 @@ angular.module('bahmni.appointments')
                     var filteredAppointments = appointmentsFilter($scope.allAppointmentsForDay, $state.params.filterParams);
                     $rootScope.appointmentsData = filteredAppointments;
                     autoRefreshStatus = true;
+                    $rootScope.appointmentStartTime = _.last($scope.allAppointmentsForDay);
```
@prashantsable1991 Two comments:

1- It should find the earliest slot, not the one available after the last appointment of the day. 
2- It should allow for appointments at the same time if they are for different services.

lauravignoli commented:
@prashantsable1991 Great work!!!

tmvumbi2 commented:
@rubailly I tried adding a new appointment from an empty calendar and got the error below. This error disappears if there exists an appointment (where the time slot has been manually populated)
![image](https://user-images.githubusercontent.com/22587098/60890189-189a4d80-a25b-11e9-9f9e-6f1d51ac1e4c.png)

tmvumbi2 commented:
@rubailly I noticed the bugs mentioned on Laura's comment.
Bug 1:
- I created an appointment on 9 July (from 10:30 to 11 am)
- I tried creating a new appointment on 10 Jul (where there is no appointment scheduled yet); I didn't specify the timeslot and the system selected 11 to 11.30 am (instead of picking the earliest slot)

Bug 2:
- I tried creating an appointment with a different service, and the system booked the next available spot. Expected result: the system should allow appointment at the same time if the service is different (or if the provider is different)

tmvumbi2 commented on ui/app/appointments/controllers/manage/appointmentsCreateController.js:
```diff
@@ -93,14 +93,49 @@ angular.module('bahmni.appointments')
                 }
             };
 
+            var setAppointmentsEndTimeAmPm = function (date) {
```
@rubailly I'd suggest we name this method `calculateAppointmentEndTime`

tmvumbi2 commented on ui/app/appointments/controllers/manage/appointmentsCreateController.js:
```diff
@@ -93,14 +93,49 @@ angular.module('bahmni.appointments')
             };
 
+            var setAppointmentsEndTimeAmPm = function (date) {
+                var strTime = moment(date).add(30, 'm').toDate();
```
@rubailly '30' in here corresponds to the appointment duration? I'm not sure if the value should be hard-coded; could you please check if there is a variable storing the minimum duration that we can use here (like `$scope.minDuration`)?

tmvumbi2 commented on ui/app/appointments/controllers/manage/appointmentsCreateController.js:
```diff
@@ -505,6 +540,7 @@ angular.module('bahmni.appointments')
             };
 
             var saveAppointment = function (appointment) {
+                console.log(appointment.startDateTime, 'starttime');
```
@rubailly please remove the debug `console.log`

tmvumbi2 committed:
COM-392 Set appointment start and end date optional

This allows the system to book the next available slot for the current date and provider and service, when the start and end date are not captured on the appointment form

rubailly commented:
@tmvumbi2 It looks like it can't find the earliest slot. I created two appointments for the same service, one at 8am and another at 2pm. I then created a third one without specifying the start time and end time. -It went to 2:30 pm 

![Screenshot from 2019-07-16 15-21-28](https://user-images.githubusercontent.com/840071/61298089-e7cb9280-a7dd-11e9-95fe-91e4d3ce581e.png)

tmvumbi2 committed:
COM-392 Book the first available slot when no time is set

This commit updates the booking logic so that the first available time slot is picked when the start and end time are not captured

rubailly commented:
![Screenshot from 2019-07-18 13-16-49](https://user-images.githubusercontent.com/840071/61453541-b1695100-a95e-11e9-8b94-bd5aeb3688c4.png)
![Screenshot from 2019-07-18 13-18-39](https://user-images.githubusercontent.com/840071/61453543-b1695100-a95e-11e9-8f67-cb54b7414d22.png)

@tmvumbi2 The way it is working now, if you don't specify the appointment start and end time, it will find the earliest appointment and use the same starting and end time. So you end up with multiple appointments that start and end at the same time for the same service and provider.

tmvumbi2 committed:
COM-392 Prevent the patient to be booked several times in same slot

This commit fixes a bug that caused the patient to be booked several times on the first time slot of the day

rubailly commented:
@tmvumbi2 -All is working well now. The remaining thing is the logic to handle the edge cases when the day has been fully booked. 1) You have to click the save button twice to see the alert that says all slots have been booked. Ideally, you should see the alert automatically after clicking the save button. 2) When you force one appointment to go on a used slot, all the other appointments you create and you don't
 specify the start/end time, they will mix with the appointment you created previously.
![Screenshot from 2019-07-19 13-50-07](https://user-images.githubusercontent.com/840071/61533397-0f646a00-aa2d-11e9-9ff2-7fb5dd04e083.png)

I, however, think that this is not in the scope of this story, I'll merge this and ping Lisa to create a separate story to handle this logic.
